### ISSUE TITLE
Navigating back on Logs page is broken

### ISSUE THREAD

**Issue by marusak** [cockpit-project/cockpit#13091]
Often going back one page with browser back button (or any browser supported shortcut) does not work as expected.
Examples: https://github.com/cockpit-project/cockpit/issues/1043 https://github.com/cockpit-project/cockpit/issues/3809

**Comment by martinpitt** [comment]
I actually think the page specific issues are better, unless they all have the same root cause?

**Comment by marusak** [comment]
> I actually think the page specific issues are better, unless they all have the same root cause?

Ah, I thought this is directly related to using `cockpit.navigation.go`

**Comment by martinpitt** [comment]
My gut feeling is that for this to work, every "state" of a page (looking at a particular service, log entry, etc.) needs to build a unique URL. I may have misunderstood, of course.

**Comment by garrett** [comment]
It's also related to jump.

@martinpitt: That's certainly part of the solution. Every page should have its own URL. We usually do this, but not always.

But even when we do this, going back doesn't always work quite right. For example, have a bunch of services on the services page, scroll down some, click on a service, then click back. You lose your place. It's already horrible for everyone, but it's even worse for screen readers.

**Comment by marusak** [comment]
Well I don't think it is only related to having different urls.
1. Go to services - url: `https://localhost:9090/system/services`
2. Go to logs - url: `https://localhost:9090/system/logs`
3. Go to specific log - url: `https://localhost:9090/system/logs#/s=<some_long_identfier>`
4. Button back - url: `https://localhost:9090/system/logs` OK
5. Button back - url:  `https://localhost:9090/system/logs#/s=<some_long_identfier>` WRONG
6. Button back - url: `https://localhost:9090/system/logs`
... and it is infinite loop between these two.

**Comment by martinpitt** [comment]
Ack, thanks for the heads-up!

**Comment by martinpitt** [comment]
This issue is too broad and too generic. Needs to be reported for particular pages/situations.

**Comment by garrett** [comment]
It seemed a lot of the issues are stemming from `cockpit.navigation.go` and `cockpit.jump` and fixing those would fix most, if not all, of the back button issues?

**Comment by martinpitt** [comment]
OK, making this issue specific to the Logs page described above.

**Comment by marusak** [comment]
Ok, so I look into this and the issue is as follows:
We have `hashchanged` event listener [here](https://github.com/cockpit-project/cockpit/blob/master/pkg/shell/base_index.js#L345) that adds the new url (with hash) into history. That means the following:
1. You are on page `/foo`, we insert `/foo` into history.
2. You navigate to `/foo#bar`, we insert `/foo#bar` into history.
3. You try to go back, but hash has changed, so this callback is again triggered, so we add `/foo` to the histroy and thus we create this infinite loop.

This problem does not happen for non-hashed pages. I believe we should not have this event listener and pages that go to different pages on hash change should be reworked not to do that - as hash is used to point on something on the page (anchor-wise) and not to navigate to different page. If we want to parametrize url we should use query parameters. Also we would like to use this hash to highlight part of a page as tracked in https://github.com/cockpit-project/cockpit/issues/14320 (which would break navigation even more).

Other solution (or rather hack) would be to keep array of visited hashed paths and when we go to same has as before it can be considered step back and we would not insert it into history.

**Comment by marusak** [comment]
oh, we use this approach on almost every single page :/ So it won't be just "stop using it" as that is gonna take some real effort to do that.

**Comment by marusak** [comment]
Uh oh, going to `https://localhost:9090/network?foo=bar` redirects it to `https://localhost:9090/network` (so effectively dropping the query string). This sucks and feels like a bug?

**Comment by garrett** [comment]
Is the problem that we're using the History API and _also_ using hash changes? Doesn't changing the hash also insert into the history? Should we only push into the history stack when not changing the hash (as doing both would cause a duplicate entry, causing a loop)?

Using the hash to specify different pages was an old hack to not get a browser to reload an SPA (Single Page App, which we kind of are sometimes) around a decade (and prior) ago. Since then, we've gotten the History API to directly change the URI to a relevant resource and also not reload the page. (The History API was literally introduced a decade ago, as https://caniuse.com/history shows: Chrome in May 2010, Firefox in March 2011.)

If we want to still programmatically change the URL, even when there's a hash, there's also `replaceState()` https://developer.mozilla.org/en-US/docs/Web/API/History/replaceState... but just using a hash should update the history anyway, so it should be unneeded. (Even when a page changes from foo#baz to bar#xyz would naturally update the history stack.)

https://css-tricks.com/using-the-html5-history-api/
https://developer.mozilla.org/en-US/docs/Web/API/History_API/Working_with_the_History_API

Agreed: We shouldn't use the hash hack, but the history API... but as you state, that would require more work. And changing it would also break current bookmarks administrators may have made, unless we did make sure to handle a translation of the URI. So I think you're right that we need to handle this in at least two different attempts. First: Unbreak current URIs. Later: Improve URIs.

**TL;DR**: Yep, agreed.

**Comment by garrett** [comment]
Every single person in the metrics UX study struggled with the back button, and some were quite frustrated. 

We absolutely need to prioritize this.

**Comment by martinpitt** [comment]
This was fixed in PR #16938 at last, thanks Marius!
